File: java/leetcode/src/Task125.java
import org.junit.jupiter.api.Test;

public class Task125 {
    public boolean isPalindrome(String s) {
        int left = 0,right = s.length()-1;

        while(left < right){
            char c1 = s.charAt(left);
            char c2 = s.charAt(right);
            if(!isValid(c1)){
                ++left;
            }else if(!isValid(c2)){
                --right;
            } else{
                if(c1 == c2|| c1 >= 'A' && c2 >= 'A' && (c1 - c2 == 32|| c1-c2 == -32 )){
                    ++left;
                    --right;
                }else {
                    return false;
                }
            }
        }

        return true;
    }

    boolean isValid(char c){
        return c >='0' && c<='9'  ||c >='a' && c<='z' ||c >='A' && c<='Z';
    }

    @Test
    void test(){
        String s = "A man, a plan, a canal: Panama";
        System.out.println(isPalindrome(s));
    }

}

File: py/Task125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True
        s = s.lower()
        left, right = 0, len(s)-1

        while left < right:
            # valid left:
            cl = s[left]
            if not cl.isdigit() and not cl.isalpha():
                left += 1
                continue
            cr = s[right]
            if not cr.isdigit() and not cr.isalpha():
                right -= 1
                continue
            if cr != cl:
                return False
            left += 1
            right -= 1
        return True


obj = Solution()
s = "A man, a plan, a canal: Panama"
print(obj.isPalindrome(s))
